### PR TITLE
[SYCL][Matrix] Add support for tf32 type using the unified interface

### PR DESCRIPTION
**Pull request by dkhaldi**

### CODE REVIEW

**Comment by JackAKirk** [comment]
LGTM.

Intel backend also does not support tf32 for Accumulator right?

btw I think this patch will also allow you to load joint_matrix from `const T`. You might want to just verify this works.

File: sycl/test/matrix/matrix-tf32-test.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+                 accB.get_pointer() + (k) * (N) + sg_starty / SG_SZ * TN, N);
+             // If no rounding to tf32 function is called, joint_matrix_mad
+             // function will work on truncated floats.
+             for (int i = 0; i < get_wi_data(sg, sub_a).length(); i++) {
+               get_wi_data(sg, sub_a)[i] =
+                   round_to_tf32(get_wi_data(sg, sub_a)[i]);
+             }
+             for (int i = 0; i < get_wi_data(sg, sub_b).length(); i++) {
+               get_wi_data(sg, sub_b)[i] =
+                   round_to_tf32(static_cast<float>(get_wi_data(sg, sub_b)[i]));
```
[the remark this answers is not in the document]
Comment (dkhaldi) [reply sycl/test/matrix/matrix-tf32-test.cpp:86]: No it is not needed I will remove it

File: sycl/test/matrix/matrix-tf32-test.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+             joint_matrix_load(
+                 sg, sub_a, accA.get_pointer() + (sg_startx * TM) * K + k, K);
+             joint_matrix_load(
+                 sg, sub_b,
+                 accB.get_pointer() + (k) * (N) + sg_starty / SG_SZ * TN, N);
+             // If no rounding to tf32 function is called, joint_matrix_mad
+             // function will work on truncated floats.
+             for (int i = 0; i < get_wi_data(sg, sub_a).length(); i++) {
+               get_wi_data(sg, sub_a)[i] =
+                   round_to_tf32(get_wi_data(sg, sub_a)[i]);
```
Comment (yubingex007-a11y) [inline sycl/test/matrix/matrix-tf32-test.cpp:82]: get_wi_data(sg, sub_a)[i] is a temporary object which can't be passed to const reference.
see https://godbolt.org/z/qncPdP8Tc.
so we should add const to round_to_tf32:
inline __SYCL_ALWAYS_INLINE float round_to_tf32( **const** float& a) {

File: sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp [hunk 3]
```diff
@@ -65,19 +65,39 @@ template <typename Group, typename T, use Use, size_t Rows, size_t Cols,
           layout Layout>
 struct joint_matrix;
 
+// Differentiating between the "element type" and the "storage element type"
```
Comment (yubingex007-a11y) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:68]: should move tf32's definition from sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcores.hpp to sycl/include/sycl/ext/oneapi/matrix/matrix-unified-utils.hpp

File: sycl/test/matrix/matrix-tf32-test.cpp (hunk 2 above)
Comment (JackAKirk) [reply sycl/test/matrix/matrix-tf32-test.cpp:82]: Looks good. Yeah should be no problems with cuda.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp (hunk 3 above)
Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:68]: yeah makes sense

**Review by MrSidims — COMMENTED** [review]
The patch is LGTM
Please hold the merge for a bit to clarify SPIR-V instruction name (Convert Vs Round)

**Pull request closed without merging**